src/network/connection.js

```diff
@@ -90,6 +90,14 @@ module.exports = class Connection {
     this.shouldLogBuffers = getEnv().KAFKAJS_DEBUG_PROTOCOL_BUFFERS === '1'
   }
 
+  get sasl() {
```

Comment:
Since `_sasl` can now be an async function, this will also need to become async, I'm afraid.

This unfortunately means that the getter approach doesn't make much sense anymore (sorry about that. Didn't think ahead). Instead, we should just make sure that `sasl` is always an async function, so something like:

```javascript
this.sasl = async () => typeof sasl === 'function' ? sasl() : sasl
```

This means that any place where we're accessing `connection.sasl` will need to change to invoke and await:

```javascript
const sasl = await connection.sasl()
```

Another option is to keep the getter, and change it to return a promise:

```javascript
get sasl() {
  return new Promise((resolve, reject) => {
    if (typeof this._sasl === 'function') {
      try {
        resolve(this._sasl())
      } catch (e) {
        reject(e)
      }
    } else {
      resolve(this._sasl)
    }
  })
}
```

Either way, all call sites would need to be updated to await the promise, so:

```javascript
const sasl = await connection.sasl
```

In my mind that feels less intuitive, so I wouldn't go with the getter.